경북대 BE_박신영 1단계 - 엔티티 매핑

parknew0 opened the pull request:
안녕하세요 멘토님! 1단계 완료했습니다.

테스트 코드는 처음 짜봐서 어떤식으로 짜면 좋을지 감이 잘 안오네요. 

잘부탁드립니다...!!

parknew0 committed:
feat: 초기 코드 마이그레이션

parknew0 committed:
feat: STEP0를 위한 초기 코드 마이그레이션

parknew0 committed:
feat(docs): readme.md 1단계 구현할 목록 업데이트

parknew0 committed:
chore(build): Spring Data JPA 의존성 추가

parknew0 committed:
chore(config): JPA 및 H2 데이터베이스 설정 추가

parknew0 committed:
refactor(product): Product 도메인 JPA로 리팩터링

parknew0 committed:
refactor(product): Member 도메인 JPA로 리팩터링

parknew0 committed:
refactor(wish): Wish 도메인 JPA로 리팩터링

parknew0 committed:
chore(docs): read.me 파일 누락된 작업 포함

parknew0 committed:
test(repository): JPA repository test 추가

leaf-upper commented on README.md:
```diff
@@ -3,3 +3,14 @@
+
+## 구현할 기능 목록 (1단계 - 엔티티 매핑)
+
+- [x] `build.gradle`에 Spring Data JPA 의존성 추가
```
👍 👍 리드미가 간결해서 보기 좋네요 리뷰할때 도움될것 같습니다. 감사합니다

leaf-upper commented on build.gradle:
```diff
@@ -19,6 +19,7 @@ repositories {
 
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
```
spring data jpa 의존성도 있는데요, spring boot starter data jpa는 spring data jpa와 뭐가 다를까요?
spring boot starter data jpa vs spring data jpa
이 2개 의존성 차이에 대해서 알아보시면 좋을것 같아요

parknew0 replied:
spring data jpa를 사용해서 repository 인터페이스를 만들면 `save()`, `findById()` 같은 기본 기능들을 직접 만들지 않아도 사용할 수 있어 편리합니다!
하지만 데이터베이스와 연결하고, 실제 SQL 명령을 실행하기 위해서 다른 도구들이 필요한데요, spring boot starter data jpa는 해당 도구들을 포함해서 가져와주기 때문에 사용자가 편리합니다!

leaf-upper replied:
@parknew0 
spring data jpa, spring boot starter jpa 둘다 말씀하신 기능은 동일 합니다.
두개의 의존성 차이는 자동구성의 여부 (AutoConfiguration) 차이가 가장 큰것 같아요.
나중에 조금더 여유가 생겨 스프링부트를 공부하실때 자동구성이 무엇인지 한번 보면 좋을것 같습니다!

leaf-upper commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -56,10 +56,7 @@ public ResponseEntity<ProductResponse> updateProduct(
-        if (!deleted) {
-            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
-        }
+        productService.deleteProduct(productId);
```
만약 productId가 음수로 들어오면 어떻게 동작할까요?

parknew0 replied:
ProductService의 deleteProduct 메서드로 -1이 전달된 다음,
서비스는 productRepository.existsById(-1)을 호출하여 ID가 -1인 상품이 데이터베이스에 있는지 확인합니다. 
상품 ID(id)는 데이터베이스에서 1부터 시작하는 Auto-increment 양수 값이므로, ID가 -1인 상품은 존재할 수 없습니다.
existsById(-1)가 false를 반환하므로, 서비스는 NoSuchElementException 예외를 발생시킬 것입니다.

데이터베이스까지 확인하러 가기 전에, Controller 계층에서 잘못된 요청임을 미리 차단하는 것이 더 효율적일 것 같아서, 그렇게 수정하겠습니다!

leaf-upper replied:
FastFail 원칙이란게 있습니다.
실패할것 같으면 빠르게 실패하는건데요~ 
말씀주신것처럼 데이터베이스 확인하러 가기전에 실패한다면 더 효율적인 애플리케이션이 될 수 있습니다.

leaf-upper commented on src/main/java/gift/entity/Member.java:
```diff
@@ -1,23 +1,32 @@
 
-    public static Member from(String email, String password) {
-        return new Member(null, email, password);
+    protected Member() {
```
protected 생성자를 쓰신 이유가 있을까요?

parknew0 replied:
JPA가 엔티티 객체를 생성할 때 리플렉션을 통해 기본 생성자를 호출해야 하므로 가장 먼저 기본 생성자를 추가했습니다. 이때 public으로 설정하면 코드 어디에서나 호출될 수 있어서 의도치 않게 불완전한 객체가 생성될 위험이 있고, private으로 설정하면 레이지 로딩 등 JPA의 프록시 기능이 동작하지 않을 수 있다고 알고있습니다. 그래서 객체 생성의 안정성을 높이기 위해 접근 제어자를 protected로 설정했습니다!

leaf-upper replied:
👍 정확하게 알고계시네요

leaf-upper commented on src/main/java/gift/entity/Member.java:
```diff
@@ -1,23 +1,32 @@
+    @Column(unique = true, nullable = false)
+    private String email;
+
+    @Column(nullable = false)
```
저는 개인적으로 `@NotNull` 애너테이션을 사용하는것도 선호하는데요,
이런방식도 괜찮은것 같아요~

leaf-upper commented on src/main/java/gift/entity/Member.java:
```diff
@@ -1,23 +1,32 @@
-    private final String email;
-    private final String password;
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
`@GeneratedValue(strategy = GenerationType.IDENTITY)` 는 어떤 기능을 하는 코드일까요?
이것말고 다른 strategy에 대해서도 알고 계실까요?

parknew0 replied:
객체를 저장할 때 id는 내가 직접 정하지 않을 테니, 데이터베이스가 알아서 번호를 매겨달라고 위임을 부탁하는 것과 같은 기능입니다! 다른 전략들도 정리해서 공부해보도록 하겠습니다!

leaf-upper replied:
👍 한번 공부해보시면 좋을것 같아요

leaf-upper commented on src/main/java/gift/entity/Member.java:
```diff
@@ -1,23 +1,32 @@
 
-    public static Member of(Long id, String email, String password) {
-        return new Member(id, email, password);
+    public Member(String email, String password) {
```
email이나 password에 대한 유효성검사가 필요할까요?
파라미터에 이메일 형식이 아닌 값이 들어오거나 패스워드가 될 수 없는값이 들어올 수 있으려나요?

parknew0 replied:
유효성 검사를 해야한다면, 엔티티 내부가 아닌 데이터가 시스템으로 들어오는 지점인 DTO에서 하는 것이 좋을 것 같습니다!
MemberRequest DTO에 검증 규칙을 추가하고, MemberController 에서 검증하는 방식으로 수정했습니다!

leaf-upper commented on src/main/java/gift/entity/Wish.java:
```diff
@@ -1,40 +1,47 @@
-    }
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "member_id", nullable = false)
+    private Member member;
```
`@ManyToOne(fetch = FetchType.LAZY)` 는 어떤 코드일까요?

parknew0 replied:
FetchType.LAZY를 사용하면 불필요한 데이터베이스 조회를 막을 수 있어 애플리케이션의 성능을 크게 향상시킬 수 있습니다!!

leaf-upper replied:
LAZY는 오히려 필요한 데이터에 대해 늦게 조회하여 여러번 쿼리가 발생하거나,
트랜잭션 범위 밖에서 데이터에 접근한다면 오류가 발생할 수도 있습니다.
항상 모든 경우에 성능을 향상 시키는것은 아니기에 어떤 경우에 해당 옵션을 사용하는지 공부하면 좋을것 같습니다.

leaf-upper commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -26,15 +27,17 @@ public MemberService(MemberRepository memberRepository, JwtTokenProvider jwtToke
         return new TokenResponse(token);
     }
 
+    @Transactional(readOnly = true)
```
트랜잭션을 readOnly로 열경우 왜 성능이 좋아질까요?

parknew0 replied:
데이터 변경 작업을 하지 않으니, 변경과 관련된 불필요한 기능들을 모두 생략해서 전달해달라는 요청이기 때문입니다!

leaf-upper replied:
네 물론 말씀주신것과 같은 부분을 데이터베이스가 고려하겠지만, 그럼 데이터베이스는 어떤 작업을 효율적으로 할 수 있을까요?
이 부분이 어렵다면, 일단 DB Isolation Level부터 공부해보시면 좋을것 같아요.

leaf-upper commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -8,41 +8,54 @@
-        return updated != null ? new ProductResponse(updated) : null;
+        Product product = productRepository.findById(id)
+                .orElseThrow(
+                        () -> new NoSuchElementException("해당 ID의 상품이 존재하지 않아 업데이트할 수 없습니다: " + id));
```
NoSuchElementException 같은 자바에서 기본으로 제공하는 예외 클래스를 사용하는것도 좋지만
프로젝트가 커질수록 애플리케이션 자체에서 정의한 예외를 사용할 필요가 있어요.
예외 관련 패키지를 만들고 거기에 본인이 예외를 정의하고, 이곳에서 예외를 던져보는건 어떨까요?

parknew0 replied:
ProductNotFoundException을 만들어 ProductService에 적용했습니다. 같은 원리에 따라 MemberNotFoundException도 추가로 생성하고 WishService에서도 전용 예외가 발생하도록 리팩토링했습니다!

leaf-upper commented on src/main/java/gift/service/WishService.java:
```diff
@@ -2,52 +2,61 @@
-        return productRepository.findAllByIdIn(productIds).stream()
-                .map(ProductResponse::new)
+        List<Wish> wishes = wishRepository.findByMember(member);
+        return wishes.stream()
```
최신자바부터는 collect를 쓰지 않고도 바로 toList()를 사용할수있어요~

parknew0 replied:
스트림(Stream)을 리스트(List)로 변환할 때, .collect(Collectors.toList()) 대신 더 짧고 직관적인 .toList() 메소드를 사용할 수 있단걸 알았습니다. 적용했습니다!

leaf-upper commented on src/test/java/gift/repository/ProductRepositoryTest.java:
```diff
@@ -0,0 +1,32 @@
+    private ProductRepository productRepository;
+
+    @Test
+    @DisplayName("상품을 저장하고 ID로 조회한다")
```
해당 테스트는 데이터베이스가 잘 동작하는지에 대한 테스트일까요?

parknew0 replied:
Product 엔티티의 JPA 매핑과 ProductRepository 인터페이스가 JPA를 통해 데이터베이스에 잘 넣고 잘 가져올 수 있는지 테스트하는 코드입니다!

leaf-upper reviewed (changes requested):
안녕하세요 신영님~
1단계 리뷰 남겨드렸습니다 몇가지 확인후에 
다시 리뷰요청 주시면 확인해볼게요!
감사합니다~

parknew0 committed:
feat(product): product ID에 validation 추가

ProductController에 @Validated를 추가하고, productId를 사용하는 코드에 대해 @Positive를 추가했다.

parknew0 committed:
Merge branch 'parknew0' into step1

parknew0 committed:
refactor(repository): 불필요한 JdbcTemplate 구현체 제거

Spring Data JPA를 사용하도록 전환함에 따라 더 이상 사용되지 않는
JdbcMemberRepository, JdbcProductRepository, JdbcWishRepository
파일을 삭제합니다.

parknew0 committed:
fix(exception): ConstraintViolationException을 처리하여 400 에러 반환

jakarta.validation.ConstraintViolationException에 대한 전용 예외 핸들러를
추가합니다.

이전에는 @Validated 컨트롤러의 유효성 검사 오류를 처리하는 핸들러가
없어서, 해당 예외가 범용 핸들러에 잡혀 부적절하게 500 에러를
반환했습니다.

parknew0 committed:
feat(member): 회원가입 및 로그인 요청에 대한 유효성 검사 추가

parknew0 committed:
refactor(exception): 전용 예외 클래스를 도입하여 예외 처리 개선

기존에 사용하던 범용적인 NoSuchElementException을 구체적인 상황을 나타내는 전용 예외 클래스로 교체하여 코드의 가독성과 유지보수성을 향상시킵니다.

- ProductNotFoundException, MemberNotFoundException 추가
- ProductService, WishService에서 ID로 엔티티 조회 실패 시, 각 상황에 맞는 전용 예외를 던지도록 수정

parknew0 committed:
refactor(stream): Stream.toList()를 사용하여 코드 간결성 및 안정성 향상

parknew0 commented:
이번 리뷰가 PR된 이후에 다음 단계 개발을 진행할 수 있는데,
2단계 리뷰요청 시간인 오후 6시까지는 한시간 보다 조금 정도밖에 시간이 남지 않아서 지금 PR이 머지되어도 시간이 촉박할 것 같습니다. 
오늘 자정까지 2단계 완료하고 리뷰 요청을 드려도 될까요??

leaf-upper reviewed (approved):
안녕하세요 신영님
코멘트 남겨주신부분 잘확인했습니다.
몇가지는 잘작성해주셨고, 조금더 보충이 필요한 부분은 코멘트로 남겨드렸습니다.
다만 해당 PR에서 코멘트를 받기는 시간이 조금 모호하여 다음 PR에서 남겨주시면
이어나가서 이야기해보면 좋을것 같아요.
고생하셨습니다 👍 👍
